packages/fxa-settings/src/lib/gql.ts

```diff
@@ -2,28 +2,71 @@
  * License, v. 2.0. If a copy of the MPL was not distributed with this
  * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
 
-import { ApolloClient, NormalizedCacheObject, from } from '@apollo/client';
+import {
+  ApolloClient,
+  ApolloLink,
+  NormalizedCacheObject,
+  from,
+} from '@apollo/client';
 import { setContext } from '@apollo/client/link/context';
 import { ErrorHandler, onError } from '@apollo/client/link/error';
 import { BatchHttpLink } from '@apollo/client/link/batch-http';
 import { cache, sessionToken, typeDefs } from './cache';
+import { GraphQLError } from 'graphql';
+import { GET_LOCAL_SIGNED_IN_STATUS } from '../components/App/gql';
+import sentryMetrics from 'fxa-shared/lib/sentry';
 
-export const pagesRequiringAuthentication = ['settings'];
+/**
+ * These operation names either require auth with a valid session token
+ * or give back a valid session token (e.g. the user was signed in).
+ * See comment above GET_LOCAL_SIGNED_IN_STATUS definition.
+ *
+ * We can improve or revisit this in FXA-7626 or FXA-7184.
+ */
+const sessionTokenOperationNames = [
+  'GetInitialMetricsState',
+  'GetInitialSettingsState',
+  'SignUp',
+];
 
-export const errorHandler: ErrorHandler = ({ graphQLErrors, networkError }) => {
-  let reauth = false;
+export const isUnauthorizedError = (error: GraphQLError) =>
+  error.message === 'Invalid token' &&
+  error.extensions?.response.error === 'Unauthorized';
+
+const afterwareLink = new ApolloLink((operation, forward) => {
+  return forward(operation).map((response) => {
+    // The error link handles GQL errors and network errors. This handles
+    // successful queries and checks to see if we should update the `isSignedIn` state.
+    const successWithAuth =
+      !response.errors &&
+      operation.query.definitions.some((definition) => {
+        return (
+          definition.kind === 'OperationDefinition' &&
+          definition.name?.value &&
+          sessionTokenOperationNames.includes(definition.name.value)
+        );
+      });
 
-  const currentPageRequiresAuthentication = pagesRequiringAuthentication.some(
-    (urlSnippet) => {
-      // We check if the url for the current page contains the path of a page which requires authentication
-      return window?.location?.pathname.includes(urlSnippet);
+    if (successWithAuth) {
+      cache.writeQuery({
+        query: GET_LOCAL_SIGNED_IN_STATUS,
+        data: { isSignedIn: true },
+      });
     }
-  );
+    return response;
+  });
+});
 
+export const errorHandler: ErrorHandler = ({ graphQLErrors, networkError }) => {
+  console.log('networkError', networkError);
```

Comment:
Stray console.log?

Reply:
Ah I swear I repushed removing this, weird.

BTW, having Sentry capture the network error here will account for [this comment](https://github.com/mozilla/fxa/pull/15493/files#diff-b2ae935037707510aa99afcf2c940f07e7aad3e25a08026d4ff9cd062fa5288cR112) in the prototype.